### PR TITLE
New plugin manager

### PR DESCRIPTION
**Pull request by ronpandolfi**
This is a complete rewrite of the plugin manager which eliminates usage of yapsy. Other improvements include:
- [x] No more cascading load failures
- [x] Significantly faster startup times
- [x] Mild improvements to UI locking
- [x] Super secret hot-reloading of plugins
- [x] No more marker files!
- [x] Plugin types are now discoverable (and extensible) by the PluginPlugin interface

Side effects include:
- [x] Widespread breaking changes to the plugin manager api

Todo:
- [x] Check that progressbar updates correctly
- [x] Considerations for live in-memory plugins
- [x] Filtered notifications (started, new, finished, updated,...?)
- [x] Configurable blacklist; SettingsPlugin to enable/disable other plugins?
- [x] attach, _notify documentation about `filter`

### CODE REVIEW

**Comment by danielballan** [comment]
Super secret, you say? 😎

File: xicam/plugins/__init__.py [hunk 1]
```diff
@@ -70,462 +72,351 @@ def load_timer():
...
+
+    def _get_entrypoint_by_name(self, name, type_name):
+        return_entrypoint = None
+        return_type = None
+        # Check all types matching type_name
+        for search_type_name in self.plugin_types.keys():
+            if type_name == search_type_name or not type_name:
+                match_entrypoint = self._entrypoints[search_type_name].get(name, None)
+                if match_entrypoint and return_entrypoint:
+                    raise NameError('Multiple plugins with the same name but different types exist. '
```
Comment (danielballan) [inline xicam/plugins/__init__.py]: The usage of `NameError` here would be confusing to me. From the docs:

> Raised when a local or global name is not found. This applies only to unqualified names. The associated value is an error message that includes the name that could not be found.

Maybe `ValueError` (if the best solution is the provide a `type_name`) or a new custom type `PluginConfigurationError` (if the the best solution is to fix the clashing packages) are worth considering.

**Review by danielballan — COMMENTED** [review]
Nice to see that entrypoints does all the things we need. To what do you attribute the speedup?

File: xicam/plugins/__init__.py [hunk 2]
```diff
@@ -70,462 +72,352 @@ def load_timer():
...
 
-class XicamPluginManager(PluginManager):
+class State(Enum):
+    READY = auto()
+    DISCOVERING = auto()
+    LOADING = auto()
+    INSTANTIATING = auto()
+
+
+class Filters(Enum):
```
Comment (birdzix) [inline xicam/plugins/__init__.py:82]: Regarding the `_notify` method, should `Filters` and/or `attach` have documentation about what a filter is used for? For example, the `filter` parameter in `attach` defines if a callback should be invoked or not (by default, filter is None, which means the callback attached will always be called).

File: xicam/plugins/__init__.py [hunk 3]
```diff
@@ -40,7 +41,8 @@
 
 from yapsy.PluginManager import NormalizePluginNameForModuleName, imp, log
 import importlib.util
-from collections import deque
+from queue import LifoQueue
```
Comment (birdzix) [inline xicam/plugins/__init__.py:44]: Why do we want a stack instead of a `Queue` here?

Comment (birdzix) [reply xicam/plugins/__init__.py:44]: after talking to @ronpandolfi, it needs to be a stack to handle plugin dependencies properly:
a -> b -> c, put(b), put(c)

**Comment by danielballan** [comment]
👏

**Pull request merged**